File: 1282/c/c.kt
fun readInt(): Int {
    return readLine()!!.toInt()
}

fun readInts(): List<Int> {
    return readLine()!!.split(" ").map(String::toInt)
}

class Solution() {
    fun solve() {
        val (n, t, a, b) = readInts()
        val problemType = readInts()
        val mandatoryTime = readInts()
        val easy = sortedMapOf<Int, Int>()
        val hard = sortedMapOf<Int, Int>()
        var totalEasy = 0
        var totalHard = 0
        for (i in 0 until n) {
            val currentEasy = easy.getOrPut(mandatoryTime[i], { 0 })
            val currentHard = hard.getOrPut(mandatoryTime[i], { 0 })
            if (problemType[i] == 0) {
                easy[mandatoryTime[i]] = currentEasy + 1
                totalEasy++
            }
            else {
                hard[mandatoryTime[i]] = currentHard + 1
                totalHard++
            }
        }
        easy.getOrPut(t + 1, { 0 })
        hard.getOrPut(t + 1, { 0 })
        var ans = 0
        var mandatoryEasy = 0
        var mandatoryHard = 0
        for (entry in easy) {
            val time = entry.key
            if (time > 0) {
                var lastTime = time - 1
                var canSolve = 0
                var timeNeeded = mandatoryEasy.toLong() * a + mandatoryHard.toLong() * b
                if (lastTime.toLong() >= timeNeeded) {
                    val remainingEasy = totalEasy - mandatoryEasy
                    val remainingHard = totalHard - mandatoryHard
                    canSolve += mandatoryEasy + mandatoryHard
                    lastTime -= timeNeeded.toInt()
                    val maxEasy = (lastTime / a).coerceAtMost(remainingEasy)
                    canSolve += maxEasy
                    lastTime -= maxEasy * a;
                    if (maxEasy == remainingEasy) {
                        val maxHard = (lastTime / b).coerceAtMost(remainingHard)
                        canSolve += maxHard
                    }
                    ans = ans.coerceAtLeast(canSolve)
                }
            }
            mandatoryEasy += easy[time]!!
            mandatoryHard += hard[time]!!
        }
        println(ans)
    }
}

fun main() {
    val t = readInt()
    for (i in 1 until t + 1) {
        val solution = Solution()
        solution.solve()
    }
}

File: 1282/c/c.py
m = int(input())
for case_num in range(m):
    n, T, a, b = map(int, input().split(' '))
    d = list(map(int, input().split(' ')))
    t = list(map(int, input().split(' ')))
    critical = set()
    for ti in t:
        critical.add(ti)
        if ti >= 1:
            critical.add(ti - 1)
    critical.add(T)
    critical = list(critical)
    critical.sort()
    k = len(critical)
    lut = dict()
    for i in range(k):
        lut[critical[i]] = i
    easy = [0 for i in range(k)]
    hard = [0 for i in range(k)]
    for i in range(n):
        idx = lut[t[i]]
        if d[i] == 0:
            easy[idx] += 1
        else:
            hard[idx] += 1
    mandatory_easy = 0
    mandatory_hard = 0
    total_hard = sum(d)
    total_easy = n - total_hard
    ans = 0
    for i in range(k):
        mandatory_easy += easy[i]
        mandatory_hard += hard[i]
        time = critical[i] - mandatory_easy * a - mandatory_hard * b
        if time < 0:
            continue
        rest_easy = total_easy - mandatory_easy
        extra_easy = min(time // a, rest_easy)
        time -= extra_easy * a
        rest_hard = total_hard - mandatory_hard
        extra_hard = min(time // b, rest_hard)
        score = mandatory_easy + mandatory_hard + extra_easy + extra_hard
        ans = max(ans, score)
    print(ans)
